libglusterfs: add function to set limit of open file descriptors

dmantipov opened the pull request:
Introduce `gf_set_nofile()` to wrap commonly (well, beyond Darwin)
used `setrlimit(RLIMIT_NOFILE, ...)`, adjust related code and drop
useless inclusion of `sys/resource.h` here and there.

Signed-off-by: Dmitry Antipov <dantipov@cloudlinux.com>
Updates: #1000

gluster-ant commented:
CLANG-FORMAT FAILURE:
Before merging the patch, this diff needs to be considered for passing clang-format

```diff --git a/libglusterfs/src/common-utils.c b/libglusterfs/src/common-utils.c
index e9eecfe1e..7ca38a837 100644
--- a/libglusterfs/src/common-utils.c
+++ b/libglusterfs/src/common-utils.c
@@ -4286,7 +4286,7 @@ gf_set_nofile(rlim_t high, rlim_t low)
 {
     int n;
     struct rlimit lim;
-    rlim_t r[2] = { high, low };
+    rlim_t r[2] = {high, low};
 
     for (n = 0; n < 2; n++)
         if (r[n] != 0) {
diff --git a/libglusterfs/src/glusterfs/libglusterfs-messages.h b/libglusterfs/src/glusterfs/libglusterfs-messages.h
index 208e58897..85ae51cfc 100644
--- a/libglusterfs/src/glusterfs/libglusterfs-messages.h
+++ b/libglusterfs/src/glusterfs/libglusterfs-messages.h
@@ -400,6 +400,6 @@ GLFS_MIG(LIBGLUSTERFS, LG_MSG_NOFILE_FAILED, "", 0)
 #define LG_MSG_STRUCT_MISS_STR "struct missing"
 #define LG_MSG_METHOD_MISS_STR "method missing(init)"
 #define LG_MSG_NOFILE_FAILED_STR                                               \
-  "failed to set maximum number of open file descriptors"
+    "failed to set maximum number of open file descriptors"
 
 #endif /* !_LG_MESSAGES_H_ */
diff --git a/xlators/features/quota/src/quota-enforcer-client.c b/xlators/features/quota/src/quota-enforcer-client.c
index 7f3b44d07..9416183ac 100644
--- a/xlators/features/quota/src/quota-enforcer-client.c
+++ b/xlators/features/quota/src/quota-enforcer-client.c
@@ -476,9 +476,10 @@ out:
     return rpc;
 }
 
-static struct rpc_clnt_procedure quota_enforcer_actors[GF_AGGREGATOR_MAXVALUE] = {
-    [GF_AGGREGATOR_NULL] = {"NULL", NULL},
-    [GF_AGGREGATOR_LOOKUP] = {"LOOKUP", NULL},
+static struct rpc_clnt_procedure quota_enforcer_actors[GF_AGGREGATOR_MAXVALUE] =
+    {
+        [GF_AGGREGATOR_NULL] = {"NULL", NULL},
+        [GF_AGGREGATOR_LOOKUP] = {"LOOKUP", NULL},
 };
 
 static struct rpc_clnt_program quota_enforcer_clnt = {
```

dmantipov commented:
/run regression

xhernandez commented on libglusterfs/src/glusterfs/libglusterfs-messages.h:
```diff
@@ -279,6 +279,7 @@ GLFS_MIG(LIBGLUSTERFS, LG_MSG_IO_URING_ENTER_FAILED, "", 0)
 GLFS_MIG(LIBGLUSTERFS, LG_MSG_IO_SYNC_TIMEOUT, "", 0)
 GLFS_MIG(LIBGLUSTERFS, LG_MSG_IO_SYNC_ABORTED, "", 0)
 GLFS_MIG(LIBGLUSTERFS, LG_MSG_IO_SYNC_COMPLETED, "", 0)
+GLFS_MIG(LIBGLUSTERFS, LG_MSG_NOFILE_FAILED, "", 0)
```
Why not this:

```suggestion
GLFS_NEW(LIBGLUSTERFS, LG_MSG_NOFILE_FAILED,
         "Failed to set maximum number of open file descriptors", 2,
         GLFS_INT64(low), GLFS_INT64(high)
)
```

dmantipov replied:
OK

xhernandez commented on libglusterfs/src/common-utils.c:
```diff
@@ -4283,3 +4280,21 @@ gf_gethostname(void)
+                return r[n];
+        }
+
+    return RLIM_INFINITY;
```
This could potentially be a valid value (i.e. there's no predefined limit). Wouldn't it be better to just return a boolean or 0/-1 instead of an `rlim_t` ?

dmantipov replied:
Voting for void return value, see below.

xhernandez commented on xlators/mgmt/glusterd/src/glusterd.c:
```diff
@@ -1553,21 +1552,9 @@ init(xlator_t *this)
-    }
-#endif
+    if (gf_set_nofile(0, 65536) == RLIM_INFINITY)
+        gf_smsg(this->name, GF_LOG_WARNING, errno, LG_MSG_NOFILE_FAILED, NULL);
```
I think it would be better to add the log message inside `gf_set_nofile()`. Callers shouldn't care to log the message.

I would also make the function definition conditional, so that on Darwin it's just defined as an empty op (`#define gf_set_nofile(_high, _low) do { } while (0)`), and as a normal function everywhere else. This way we can call it without needing to add the `#ifdef` block each time.

dmantipov replied:
OK so if we don't need a special return value to indicate error (and so issue warning), it looks reasonable to convert `gf_set_nofile()` to plain `void`.

Commit:
libglusterfs: add function to set limit of open file descriptors

Introduce 'gf_set_nofile()' to wrap commonly (well, beyond Darwin)
used 'setrlimit(RLIMIT_NOFILE, ...)', adjust related code and drop
useless inclusion of 'sys/resource.h' here and there.

Signed-off-by: Dmitry Antipov <dantipov@cloudlinux.com>
Updates: #1000

gluster-ant commented:
CLANG-FORMAT FAILURE:
Before merging the patch, this diff needs to be considered for passing clang-format

```diff --git a/libglusterfs/src/common-utils.c b/libglusterfs/src/common-utils.c
index 9afa117d4..228371113 100644
--- a/libglusterfs/src/common-utils.c
+++ b/libglusterfs/src/common-utils.c
@@ -4288,7 +4288,7 @@ gf_set_nofile(rlim_t high, rlim_t low)
 {
     int n, ret = -1;
     struct rlimit lim;
-    rlim_t r[2] = { high, low };
+    rlim_t r[2] = {high, low};
 
     for (n = 0; n < 2; n++)
         if (r[n] != 0) {
diff --git a/libglusterfs/src/glusterfs/common-utils.h b/libglusterfs/src/glusterfs/common-utils.h
index 6395b5a37..b9a6c7a78 100644
--- a/libglusterfs/src/glusterfs/common-utils.h
+++ b/libglusterfs/src/glusterfs/common-utils.h
@@ -209,7 +209,9 @@ gf_set_nofile(rlim_t high, rlim_t low);
 
 #else /* Darwin */
 
-#define gf_set_nofile(low, high) do { } while (0)
+#define gf_set_nofile(low, high)                                               \
+    do {                                                                       \
+    } while (0)
 
 #endif /* not GF_DARWIN_HOST_OS */
 
diff --git a/libglusterfs/src/glusterfs/libglusterfs-messages.h b/libglusterfs/src/glusterfs/libglusterfs-messages.h
index 751dd41fd..ca39d4538 100644
--- a/libglusterfs/src/glusterfs/libglusterfs-messages.h
+++ b/libglusterfs/src/glusterfs/libglusterfs-messages.h
@@ -267,28 +267,17 @@ GLFS_MIG(LIBGLUSTERFS, LG_MSG_INPUT_DATA_NULL, "", 0)
 GLFS_MIG(LIBGLUSTERFS, LG_MSG_OPEN_LOGFILE_FAILED, "", 0)
 
 GLFS_NEW(LIBGLUSTERFS, LG_MSG_IO_CALL_FAILED, "Function call failed", 2,
-    GLFS_FUNC(function),
-    GLFS_RES(error)
-)
+         GLFS_FUNC(function), GLFS_RES(error))
 
 GLFS_NEW(LIBGLUSTERFS, LG_MSG_IO_THREAD_BAD_PRIORITY,
-         "Specified priority is out of bounds", 1,
-    GLFS_I32(priority)
-)
+         "Specified priority is out of bounds", 1, GLFS_I32(priority))
 
 GLFS_NEW(LIBGLUSTERFS, LG_MSG_IO_CBK_SLOW,
-         "Execution of callback took too much time", 5,
-    GLFS_FUNC(name),
-    GLFS_PTR(address),
-    GLFS_STR(file),
-    GLFS_U32(line),
-    GLFS_U64(us)
-)
+         "Execution of callback took too much time", 5, GLFS_FUNC(name),
+         GLFS_PTR(address), GLFS_STR(file), GLFS_U32(line), GLFS_U64(us))
 
 GLFS_NEW(LIBGLUSTERFS, LG_MSG_IO_THREAD_NO_CPU,
-         "Cannot find a suitable CPU for a thread", 1,
-    GLFS_U32(index)
-)
+         "Cannot find a suitable CPU for a thread", 1, GLFS_U32(index))
 
 GLFS_NEW(LIBGLUSTERFS, LG_MSG_IO_THREAD_NAME_INVALID,
          "Tried to construct and invalid name for a thread", 0)
@@ -306,43 +295,32 @@ GLFS_NEW(LIBGLUSTERFS, LG_MSG_IO_URING_MISSING_FEAT,
          0)
 
 GLFS_NEW(LIBGLUSTERFS, LG_MSG_IO_URING_TOO_SMALL,
-         "Maximum allowed SQ size is too small", 1,
-    GLFS_U32(max)
-)
+         "Maximum allowed SQ size is too small", 1, GLFS_U32(max))
 
 GLFS_NEW(LIBGLUSTERFS, LG_MSG_IO_URING_ENTER_FAILED,
          "io_uring_enter() failed with an unrecoverable error. This could "
          "mean a critical bug or a memory corruption. The process cannot "
-         "continue in this state", 1,
-    GLFS_RES(error)
-)
+         "continue in this state",
+         1, GLFS_RES(error))
 
 GLFS_NEW(LIBGLUSTERFS, LG_MSG_IO_SYNC_TIMEOUT,
-         "Time out while waiting for synchronization", 1,
-    GLFS_U32(retries)
-)
+         "Time out while waiting for synchronization", 1, GLFS_U32(retries))
 
 GLFS_NEW(LIBGLUSTERFS, LG_MSG_IO_SYNC_ABORTED,
-         "Synchronization took too much time", 1,
-    GLFS_U32(retries)
-)
+         "Synchronization took too much time", 1, GLFS_U32(retries))
 
 GLFS_NEW(LIBGLUSTERFS, LG_MSG_IO_SYNC_COMPLETED, "Synchronization completed", 1,
-    GLFS_U32(retries)
-)
+         GLFS_U32(retries))
 
 GLFS_NEW(LIBGLUSTERFS, LG_MSG_IO_BAD_ERRNO, "Unexpected errno value", 1,
-    GLFS_I32(err)
-)
+         GLFS_I32(err))
 
 GLFS_NEW(LIBGLUSTERFS, LG_MSG_IO_BAD_RETURN, "Unexpected return value", 1,
-    GLFS_I32(ret)
-)
+         GLFS_I32(ret))
 
 GLFS_NEW(LIBGLUSTERFS, LG_MSG_NOFILE_FAILED,
-    "Failed to set maximum number of open file descriptors", 2,
-    GLFS_I64(low), GLFS_I64(high)
-)
+         "Failed to set maximum number of open file descriptors", 2,
+         GLFS_I64(low), GLFS_I64(high))
 
 #define LG_MSG_EPOLL_FD_CREATE_FAILED_STR "epoll fd creation failed"
 #define LG_MSG_INVALID_POLL_IN_STR "invalid poll_in value"
diff --git a/xlators/features/quota/src/quota-enforcer-client.c b/xlators/features/quota/src/quota-enforcer-client.c
index 7f3b44d07..9416183ac 100644
--- a/xlators/features/quota/src/quota-enforcer-client.c
+++ b/xlators/features/quota/src/quota-enforcer-client.c
@@ -476,9 +476,10 @@ out:
     return rpc;
 }
 
-static struct rpc_clnt_procedure quota_enforcer_actors[GF_AGGREGATOR_MAXVALUE] = {
-    [GF_AGGREGATOR_NULL] = {"NULL", NULL},
-    [GF_AGGREGATOR_LOOKUP] = {"LOOKUP", NULL},
+static struct rpc_clnt_procedure quota_enforcer_actors[GF_AGGREGATOR_MAXVALUE] =
+    {
+        [GF_AGGREGATOR_NULL] = {"NULL", NULL},
+        [GF_AGGREGATOR_LOOKUP] = {"LOOKUP", NULL},
 };
 
 static struct rpc_clnt_program quota_enforcer_clnt = {
diff --git a/xlators/storage/posix/src/posix-inode-fd-ops.c b/xlators/storage/posix/src/posix-inode-fd-ops.c
index c9e6a3423..2fded7375 100644
--- a/xlators/storage/posix/src/posix-inode-fd-ops.c
+++ b/xlators/storage/posix/src/posix-inode-fd-ops.c
@@ -5972,10 +5972,7 @@ posix_readdirp(call_frame_t *frame, xlator_t *this, fd_t *fd, size_t size,
         if (op_ret >= 0) {
             op_ret = 0;
 
-            list_for_each_entry(entry, &entries.list, list)
-            {
-                op_ret++;
-            }
+            list_for_each_entry(entry, &entries.list, list) { op_ret++; }
         }
 
         STACK_UNWIND_STRICT(readdirp, frame, op_ret, op_errno, &entries, NULL);
```

dmantipov commented:
/run regression
